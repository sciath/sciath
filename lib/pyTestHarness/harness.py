import os,sys
import argparse
import pyTestHarness.unittest as unittest
import pyTestHarness.launch as launch
from   pyTestHarness.colors import pthNamedColors as bcolors


class pthHarness:
  def __init__(self,registeredTests):
    self.testsRegistered = 0
    self.testsExecuted = 0
    self.testsSkipped = 0
    self.testsPassed = 0
    self.testsFailed = 0
    self.verbosity_level = 1
    self.pthErrorReportFileName = 'pthErrorReport.log'

    self.testDescription = []
    self.registeredTests = registeredTests

    for t in self.registeredTests:
      if not isinstance(t,unittest.pthUnitTest):
        raise ValueError('[pth]: Registered tests must be of type UnitTest')
    self.testsRegistered = len(self.registeredTests)

    parser = argparse.ArgumentParser(description='Python Test Harness.')
    parser.add_argument('-e', '--execute', help='Execute all tests', required=False, action='store_true')
    parser.add_argument('-v', '--verify', help='Perform test verification only (and not execution)', required=False, action='store_true')
    parser.add_argument('-c', '--configure', help='Configure queuing system information', required=False, action='store_true')
    parser.add_argument('-t', '--test', help='List of test names', required=False)
    parser.add_argument('-o', '--output_path', help='Directory to write stdout into', required=False)
    parser.add_argument('-p', '--purge_output', help='Delete generated output', required=False, action='store_true')
    parser.add_argument('-f', '--error_on_test_failure', help='Return exit code of 1 if any test failed', required=False, action='store_true')
    parser.add_argument('-d', '--configure_default', help='Write default queuing system config file (no mpi, no queuing system)', required=False, action='store_true')
    parser.add_argument('-s', '--sandbox', help='Execute tests in separate directories. Will not work unless you supply absolute paths to executables.', required=False, action='store_true')
    parser.add_argument('-l', '--list', help='List all registered tests and exit', required=False, action='store_true')
    self.args, self.unknown = parser.parse_known_args()

    # If "list" option supplied, simply print out all the tests' names and exit
    if self.args.list :
      for test in registeredTests :
        print(test.name)
      sys.exit(0)

    # If --configure_default is specified, write the default file and exit
    if self.args.configure_default:
      launch.pthLaunch.writeDefaultDefinition()
      sys.exit(0)

    # Create the launcher
    self.launcher = launch.pthLaunch()

    # If --configure is specified, (re)generate the configuration file and exit
    if self.args.configure:
      self.launcher.configure()
      sys.exit(0)

    # Instruct tests to use sandboxes if requested
    if self.args.sandbox :
      for test in registeredTests :
        test.use_sandbox = True

    # Set output path on all tests if --output_path option was included
    if self.args.output_path:
      for test in self.registeredTests:
        test.setOutputPath(self.args.output_path)

    # Label tests as Registered or Excluded:Reason
    for i in range(0,len(self.registeredTests)):
      self.testDescription.append('Registered')

    # Exclude parallel tests if mpiLauncher = 'none' and test uses more than 1 rank
    counter = 0
    skipCounter = 0
    for t in self.registeredTests:
      if self.launcher.mpiLaunch == 'none' and t.ranks != 1:
        self.registeredTests[counter].ignore = True
        self.testDescription[counter] = 'No MPI launcher was provided and test requested > 1 MPI rank'
        skipCounter = skipCounter + 1
      counter = counter + 1

    # Exclude tests based on command line option
    subList = []
    if self.args.test:
      tnames = self.args.test.split(',')
      for name in tnames:
        found = False
        for t in self.registeredTests:
          if name == t.name:
            subList.append(t)
            found = True
            break

        if found == False:
          errstr= '[pth] You requested to test a subset of registered tests, \n\t\t  but no registered test matched the name \"' + name + '\"'
          raise RuntimeError(errstr)

    if self.args.test:
      counter = 0
      for t in self.registeredTests:
        # skip tests already marked to be ignored (e.g. due to mpiLaunch = none and test.ranks != 1
        if t.ignore == True:
          counter = counter + 1
          continue

        if t not in subList:
          t.ignore = True
          self.testDescription[counter] = 'Excluded based on users command line arg -t'
        counter = counter + 1

    # Clean all output unless we are verifying (only)
    if not self.args.verify :
      self.clean()

    # If -p was supplied, stop after cleaning
    if self.args.verify and self.args.purge_output :
      raise RuntimeError('-p (--purge_output) and -v (--verify) are not supported together')
    if self.args.purge_output :
      exit(0);

  def setVerbosityLevel(self,value):
    self.verbosity_level = value
    self.launcher.setVerbosityLevel(self.verbosity_level)

  def execute(self):
    '''Execute tests unless verifying (only) or purging output (only)'''
    if not self.args.verify and not self.args.purge_output:
      print('')
      print(bcolors.HEADER + '[ *** Executing Tests *** ]' + bcolors.ENDC)
      launcher = self.launcher
      testList = self.registeredTests
      description = self.testDescription
      if launcher.verbosity_level > 0:
        launcher.view()

      skipCounter = 0
      for t in testList:
        if launcher.mpiLaunch == 'none' and t.ranks != 1:
          if t.ignore == True:
            skipCounter = skipCounter + 1

      if skipCounter != 0:
        print('\n' + bcolors.WARNING + 'Warning: ' + str(skipCounter) + ' MPI parallel jobs are being skipped as a valid MPI launcher was not provided'+ bcolors.ENDC)

      counter = 0
      for test in testList:
        if test.ignore == False:
          print('[-- Executing test: ' + test.name + ' --]')
          launcher.submitJob(test)
        else:
          print('[-- Skipping test: ' + test.name + ' --] Reason: ' + description[counter])
        counter = counter + 1

  def verify(self):
    '''Verify, unless we are running with a batch system and are not in verify(-only) mode'''
    if not self.launcher.useBatch or self.args.verify :
      print('')
      print(bcolors.HEADER + '[ *** Verifying Test Output *** ]' + bcolors.ENDC)
      tests_not_skipped = 0
      for test in self.registeredTests:
        print('[-- Verifying test: ' + test.name + ' --]')
        if test.ignore :
          print('[Skipping verification for test \"' + test.name + '\"]')
        elif self.launcher.mpiLaunch == 'none' and test.ranks != 1:
          print('[Skipping verification for test \"' + test.name + '\" as test uses > 1 MPI ranks and no MPI launcher was provided]')
        else:
          tests_not_skipped = tests_not_skipped + 1
          test.verifyOutput()

      print('')
      counter = 0
      for test in self.registeredTests:
        if not test.ignore and not test.passed :
          counter = counter + 1
      if counter > 0:
        print('')
        print('[--------- Unit test error report ----------------------]')
        for test in self.registeredTests:
          test.report('log')

      errfile = self.reportAll()
      if errfile and self.args.error_on_test_failure:
        print('\n')
        print('Contents of "' + errfile +'"')
        os.system('cat ' + errfile)
        sys.exit(1)

  def clean(self):
    print('')
    print(bcolors.HEADER + '[ *** Deleting Existing Test Output *** ]' + bcolors.ENDC)
    if os.path.isfile(self.pthErrorReportFileName) :
      os.remove(self.pthErrorReportFileName)
    for test in self.registeredTests:
      if test.ignore:
        print('[ -- Skipping test:',test.name,'--]')
      else:
        self.launcher.clean(test)

  def reportAll(self):
    launcher = self.launcher
    testList = self.registeredTests
    print('')
    nTests = len(testList)
    failCounter = 0
    execCounter = 0
    skipCounter = 0
    seqCounter = 0
    mpiCounter  = 0
    seqPassedCounter = 0
    mpiPassedCounter  = 0
    seqExecCounter = 0
    mpiExecCounter  = 0

    for test in testList:
      if test.ranks == 1:
        seqCounter = seqCounter + 1
      else:
        mpiCounter = mpiCounter + 1

      if test.ignore == False:
        execCounter = execCounter + 1

        if test.ranks == 1:
          seqExecCounter = seqExecCounter + 1
        else:
          mpiExecCounter = mpiExecCounter + 1

        if test.passed == False:
          failCounter = failCounter + 1
      else:
        skipCounter = skipCounter + 1

    for test in testList:
      if test.ignore == False:
        if test.ranks == 1 and test.passed == True:
          seqPassedCounter = seqPassedCounter + 1
        elif test.ranks >= 1 and test.passed == True:
          mpiPassedCounter = mpiPassedCounter + 1

    print('[--------- UnitTest status ----------------------]')
    for test in testList:
      test.report('summary')

    print('')
    print('[--------- UnitTest report ----------------------]')
    print('  ' + ("%.4d" % nTests) + ' ' + 'UnitTests registered')
    print('  ' + ("%.4d" % seqCounter) + ' Sequential UnitTests')
    print('  ' + ("%.4d" % mpiCounter) + ' MPI UnitTests')
    print('  ' + ("%.4d" % execCounter) + ' of ' +("%.4d" % nTests)+ ' UnitTests executed')

    print('')
    if execCounter == 0:
      print(bcolors.WARNING + ' [status] UNKNOWN: All tests were skipped' + bcolors.ENDC)

    elif failCounter > 0:
      print(bcolors.FAIL + ' [status] FAIL: ' + str(failCounter) + ' of ' + str(execCounter) + bcolors.FAIL + ' tests executed FAILED' + bcolors.ENDC)

      print(bcolors.FAIL +'          ' + ("%.4d" % (seqExecCounter-seqPassedCounter)) + ' of ' +("%.4d" % seqExecCounter)+ ' executed Sequential tests failed'+bcolors.ENDC)
      print(bcolors.FAIL +'          ' + ("%.4d" % (mpiExecCounter-mpiPassedCounter)) + ' of ' +("%.4d" % mpiExecCounter)+ ' executed MPI tests failed'+bcolors.ENDC)
    else:
      if skipCounter == 0:
        print(bcolors.OKGREEN + ' [status] SUCCESS: All registered tests passed' + bcolors.ENDC)
      else:
        print(bcolors.WARNING + ' [status] SUCCESS (partial): All executed tests passed' + bcolors.ENDC)

    if seqExecCounter + mpiExecCounter != nTests:
      print(bcolors.WARNING+'          Warning: Not all UnitTests were executed!'+ bcolors.ENDC)
    if seqExecCounter != seqCounter:
      print(bcolors.WARNING+'          Warning: '+("%.4d" % (seqCounter-seqExecCounter))+' sequential UnitTests were skipped'+ bcolors.ENDC)
    if mpiExecCounter != mpiCounter:
      print(bcolors.WARNING+'          Warning: '+("%.4d" % (mpiCounter-mpiExecCounter))+' MPI UnitTests were skipped!'+ bcolors.ENDC)

    errfile = []
    if failCounter > 0:
      file = open(self.pthErrorReportFileName,'w')
      sys.stdout = file

      for test in testList:
        test.report('log')

      file.close()
      sys.stdout = sys.__stdout__

      print('xxx============================================================================xxx')
      print('     UnitTests failed - Full error report written to ' + self.pthErrorReportFileName)
      print('                      - Inspect the error log file and resolve failed tests')
      pthErrorReportFileLocation = os.path.realpath(self.pthErrorReportFileName)
      print('     cat ' + pthErrorReportFileLocation)
      print('xxx============================================================================xxx')
      errfile = pthErrorReportFileLocation
    return(errfile)
