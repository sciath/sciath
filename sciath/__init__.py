""" SciATH: Scientific Application Test Harness """
from sciath._sciath_io import NamedColors

__version__ = (0, 7, 0)

# A default set of colors
SCIATH_COLORS = NamedColors()
