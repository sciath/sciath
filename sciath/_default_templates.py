""" Generate default templates for Launcher batch systems """

import os

CONTENTS_LOCAL = r"""#!/usr/bin/env sh

$SCIATH_TASK_MPI_RUN \
$SCIATH_TASK_COMMAND 1>>$SCIATH_JOB_STDOUT 2>>$SCIATH_JOB_STDERR; printf "$?\n" >> $SCIATH_JOB_EXITCODE

touch $SCIATH_JOB_COMPLETE
"""

CONTENTS_LSF = r"""#!/bin/sh

#BSUB -J $SCIATH_JOB_NAME
#BSUB -o $SCIATH_JOB_STDOUT
#BSUB -e $SCIATH_JOB_STDERR
#BSUB -n $SCIATH_JOB_MAX_RANKS_OR_REMOVE_LINE
#BSUB -W $SCIATH_JOB_WALLTIME_HM_OR_REMOVE_LINE
#BSUB -q $SCIATH_QUEUE_OR_REMOVE_LINE

$SCIATH_TASK_MPI_RUN \
$SCIATH_TASK_COMMAND; printf "$?\n" >> $SCIATH_JOB_EXITCODE

touch $SCIATH_JOB_COMPLETE
"""

CONTENTS_SLURM = r"""#!/bin/bash -l

#SBATCH --job-name=$SCIATH_JOB_NAME
#SBATCH --output=$SCIATH_JOB_STDOUT
#SBATCH --error=$SCIATH_JOB_STDERR
#SBATCH --ntasks=$SCIATH_JOB_MAX_RANKS_OR_REMOVE_LINE
#SBATCH --time=$SCIATH_JOB_WALLTIME_HMS_OR_REMOVE_LINE
#SBATCH --account=$SCIATH_ACCOUNT_OR_REMOVE_LINE
#SBATCH --partition=$SCIATH_QUEUE_OR_REMOVE_LINE

$SCIATH_TASK_MPI_RUN \
$SCIATH_TASK_COMMAND; printf "$?\n" >> $SCIATH_JOB_EXITCODE

touch $SCIATH_JOB_COMPLETE
"""


def _generate_default_template(system_type, output_path=".", filename=None):
    if system_type == "local":
        contents = CONTENTS_LOCAL
        if filename is None:
            filename = "SciATH_template_local.sh"
    elif system_type == "lsf":
        contents = CONTENTS_LSF
        if filename is None:
            filename = "SciATH_template_lsf.lsf"
    elif system_type == "slurm":
        contents = CONTENTS_SLURM
        if filename is None:
            filename = "SciATH_template_slurm.sbatch"
    else:
        raise Exception("Unrecognized batch system type %s" % system_type)

    with open(os.path.join(output_path, filename), "w") as file:
        file.write(contents)
    return filename
