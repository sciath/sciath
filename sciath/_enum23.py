# Defines a class which behaves like an Enum (Python 3.4+)
# But also supports some operations with earlier versions of Python
# It is used just like Enum, except that values must be assigned
# Using Enum23Value(), not directly, for example
#
# class MyEnum(Enum23):
#    FOO = Enum23Value('foo')
#
# If Python <3.4 support is dropped, Enum23 can be directly replaced
# with Enum, and Enum23Value() can be removed

import sys

if sys.version_info[0] >= 3 and sys.version_info[1] >= 4:
    from enum import Enum

    def Enum23Value(x):
        return x

    Enum23 = Enum
else:
    class Enum23Value:
        """ A wrapper class which allows one to access a value field """
        def __init__(self, value): self.__dict__.update(value=value)

    Enum23 = object
