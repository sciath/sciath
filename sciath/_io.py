from __future__ import print_function

import os
import sys

def py23input(prompt) :
    if sys.version_info[0] == 2 :
        v = raw_input(prompt)
    else :
        v = input(prompt)
    return(v)


def _remove_file_if_it_exists(filename):
    try:
        os.remove(filename)
    except OSError:
        pass


def command_join(command):
    """ Convert a command (as would go to subprocess.run()) to a copy-pasteable string

    Do something similar to shlex.join (Python 3.8+), attempting to quote arguments
    that contain spaces, and escape newlines in the result. """

    joined = ' '.join(["'"+term+"'" if ' ' in term else term for term in command])
    joined = joined.replace('\n','\\n')
    return joined

class NamedColors:
    def __init__(self):
        self.set_colors()

    def set_colors(self,use_bash=True):
        self.HEADER    = '\033[35m' if use_bash else ''
        self.SUBHEADER = '\033[36m' if use_bash else ''
        self.OK        = '\033[32m' if use_bash else ''
        self.WARNING   = '\033[93m' if use_bash else ''
        self.FAIL      = '\033[91m' if use_bash else ''
        self.ENDC      = '\033[0m'  if use_bash else ''

# two space tab for formatted print statements
tab = '  '
