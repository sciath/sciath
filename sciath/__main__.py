import sciath.harness

def main():
    sciath.harness.Harness().run_from_args()

if __name__ == '__main__':
    main()
