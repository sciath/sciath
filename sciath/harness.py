from __future__ import print_function

import os
import sys
import argparse
import shutil

import sciath
import sciath.launcher
import sciath._test_file
from sciath import sciath_colors
from sciath._sciath_io import py23input

class _TestRunStatus:
    DEACTIVATED             = 'deactivated'  # Test skipped intentionally
    UNKNOWN                 = 'unknown'      # Neither checked for completion nor verified
    NOT_LAUNCHED            = 'not launched' # Launcher reports test has not been launched
    INCOMPLETE              = 'incomplete'   # Launcher reports test run incomplete
    COMPLETE_AND_UNVERIFIED = 'unverified'   # Launcher reports complete, yet verification not performed
    SKIPPED                 = 'skipped'      # Test skipped because Launcher reports of lack of resources
    JOB_INVALID             = 'invalid job'  # Launcher reports test's Job is incorrectly specified
    TEST_INVALID            = 'invalid test' # Verifier reports test/verification is badly specified
    PASS                    = 'pass'         # Verifier confirms pass
    FAIL                    = 'fail'         # Verifier confirms fail

class _TestRun:
        """ A private class which adds state about a specific "run" of a Test.

        It contains a Test object, which should be thought of as the stateless
        information about a test case, provided by a user. In addition, it
        contains information managed by the Harness, such as a location to run
        from, information collected from the Launcher (to use for output), etc.
        """

        def __init__(self,test):
            self.active = True
            self.test = test
            self.output_path = os.path.join(os.getcwd(),test.name + '_output')
            self.exec_path = os.path.join(self.output_path,'sandbox')
            self.sandbox = True
            self.status = _TestRunStatus.UNKNOWN
            self.status_info = ''
            self.report = []

class Harness:
    """ :class:`Harness` is the central user-facing class in SciATH.

    It manages a set of tests, and thus includes:

    * A set of uniquely-named :class:`Test` objects
    * A :class:`Launcher`
    * Tools for running and verifying a test suite
    * Tools for managing "sandboxing" (managing directories in which to run tests)


    It is the exclusive location within SciATH for

    * Printing to stdout
    * Information about where to launch :class:`Job`s from, passed to included `Launcher`s

    A :class:`Harness` object's state is confined to the state of a list of internal
    :class:`_TestRun` objects.
    """

    _sandbox_sentinel_filename = '.sciath_sandbox'

    def __init__(self, tests=[]):
        self.launcher = None # Created when needed
        self.testruns = []
        for test in tests:
            self.add_test(test)

    def add_test(self, test):
        if test.name in [testrun.test.name for testrun in self.testruns]:
            raise Exception("Duplicate test name %s" % test.name)
        self.testruns.append(_TestRun(test))

    def add_tests_from_file(self, filename):
        for test in sciath._test_file.create_tests_from_file(filename):
            self.add_test(test)

    def clean(self):
        if self.launcher is None:
            self.launcher = sciath.launcher.Launcher()

        # Clean all tests
        if self.testruns:
            print(sciath_colors.HEADER+'[ *** Cleanup *** ]'+sciath_colors.ENDC)
        for testrun in self.testruns:
            if testrun.active:
                print('[ -- Removing output for Test:',testrun.test.name,'-- ]')
                self.launcher.clean(testrun.test.job, output_path=testrun.output_path)
                if testrun.sandbox and os.path.exists(testrun.exec_path):
                    sentinel_file = os.path.join(testrun.exec_path,self._sandbox_sentinel_filename)
                    if not os.path.exists(sentinel_file):
                        raise Exception('[SciATH] did not find expected sentinel file ' + sentinel_file)
                    shutil.rmtree(testrun.exec_path)

    def determine_overall_success(self):
        for testrun in self.testruns:
            if testrun.status not in [_TestRunStatus.DEACTIVATED, _TestRunStatus.PASS] :
                return False
        return True

    def execute(self):
        self.clean()

        if self.launcher is None:
            self.launcher = sciath.launcher.Launcher()

        if self.testruns:
            print(sciath_colors.HEADER + '[ *** Executing Tests *** ]' + sciath_colors.ENDC)
            self.launcher.view()
        for testrun in self.testruns:
            if testrun.active:
                if not os.path.exists(testrun.output_path):
                    os.makedirs(testrun.output_path)
                if not os.path.exists(testrun.exec_path):
                    os.makedirs(testrun.exec_path)
                if testrun.sandbox:
                    sentinel_file = os.path.join(testrun.exec_path,self._sandbox_sentinel_filename)
                    if os.path.exists(sentinel_file):
                        raise Exception("[SciATH] Unexpected sentinel file " + sentinel_file)
                    with open(sentinel_file,'w'):
                        pass
                self.launcher.submitJob(
                        testrun.test.job,
                        output_path=testrun.output_path,
                        exec_path = testrun.exec_path)

    def print_all_tests(self):
        for testrun in self.testruns:
            print(testrun.test.name)

    def report(self):
        """ Compile results into a report and print """
        failed_names = []
        if self.testruns:
            report_header_printed = False
            for testrun in self.testruns:
                if testrun.report:
                    if not report_header_printed:
                        print(sciath_colors.HEADER+'[ *** Verification Reports *** ]'+sciath_colors.ENDC)
                        report_header_printed = True
                    print('%s[Report for %s]%s' % (sciath_colors.SUBHEADER,testrun.test.name,sciath_colors.ENDC))
                    for line in testrun.report:
                        print(line)
            print(sciath_colors.HEADER+'[ *** Summary *** ]'+sciath_colors.ENDC)
            for testrun in self.testruns:
                if testrun.status == 'fail':
                    failed_names.append(testrun.test.name)
                print(sciath.sciath_test_status.status_color_type[testrun.status], end='')
                print('[%s]  %s' %(testrun.test.name, testrun.status), end='')
                print(sciath_colors.ENDC, end='')
                if testrun.status_info:
                    print(' (' + testrun.status_info + ')', end='')
                print()
            print()
            if self.determine_overall_success():
                print(sciath_colors.OK + "SUCCESS" + sciath_colors.ENDC)
            else:
                print(sciath_colors.FAIL + "FAILURE" + sciath_colors.ENDC)
                if failed_names:
                    print('To re-run failed tests, use e.g.')
                    print('  -t ' + ','.join(failed_names))
        else:
            print("No tests")

    def run_from_args(self):
        """ Perform one or more actions, based on command line options

        This essentially defines the "main" function for the typical
        use of SciATH.
        """

        parser = argparse.ArgumentParser(description='SciATH')
        parser.add_argument('-c', '--configure', help='Configure queuing system information', required=False, action='store_true')
        parser.add_argument('-t', '--test-subset', help='Comma-separated list of test names', required=False)
        parser.add_argument('-p', '--purge-output', help='Delete generated output', required=False, action='store_true')
        parser.add_argument('-f', '--error-on-test-failure', help='Return exit code of 1 if any test failed', required=False, action='store_true')
        parser.add_argument('-d', '--configure-default', help='Write default queuing system config file (no mpi, no queuing system)', required=False, action='store_true')
        parser.add_argument('-l', '--list', help='List all registered tests and exit', required=False, action='store_true')
        parser.add_argument('-w','--conf-file',help='Use provided configuration file instead of the default',required=False)
        parser.add_argument('--no-colors',help='Deactivate colored output',required=False,action='store_true')
        parser.add_argument('-i', '--input-file', help='Parse a file to add tests to the harness', required=False)
        parser.add_argument('-u', '--update-expected', help='When well-defined, update reference files with current output before verifying', required=False, action='store_true')
        stage_skip_group = parser.add_mutually_exclusive_group()
        stage_skip_group.add_argument('-v', '--verify', help='Perform test verification, and not execution', required=False, action='store_true')
        stage_skip_group.add_argument('-e','--execute', help='Perform test execution, and not verification', required=False, action='store_true')
        args,unknown = parser.parse_known_args()

        if args.no_colors:
            sciath.sciath_colors.set_colors(use_bash = False)

        if args.update_expected:
            print("[SciATH] You have provided an argument to updated expected files.")
            print("[SciATH] This will attempt to OVERWRITE your expected files!")
            if py23input("[SciATH] Are you sure? Type 'y' to continue: ")[0] not in ['y','Y']:
                print("[SciATH] Aborting.")
                return

        if args.input_file:
            self.add_tests_from_file(args.input_file)

        if args.list:
            self.print_all_tests()
            return

        if args.test_subset:
            self._activate_tests_from_argument(args.test_subset)

        if args.configure_default:
            sciath.launcher.Launcher.writeDefaultDefinition(args.conf_file)
            return

        self.launcher = sciath.launcher.Launcher(args.conf_file)

        if args.configure:
            self.launcher.configure()

        if args.purge_output:
            self.clean()
            return

        if not args.verify:
            self.execute()

        if args.update_expected:
            self.update_expected()

        if args.execute or (not args.verify and self.launcher.useBatch):
            if self.testruns:
                print('[SciATH] Not verifying or reporting')
        else:
            self.verify()
            self.report()

        if args.error_on_test_failure:
            if not self.determine_overall_success():
                sys.exit(1)

    def update_expected(self):
        """ Give each active test the chance to update its reference output """
        if self.testruns:
            print(sciath_colors.HEADER + '[ *** Updating Expected Output *** ]' + sciath_colors.ENDC)
        for testrun in self.testruns:
            if testrun.active:
                if hasattr(testrun.test.verifier, 'update_expected'):
                    print('[ -- Updating output for Test:',testrun.test.name,'-- ]')
                    testrun.test.verifier.update_expected(testrun.output_path, testrun.exec_path)
                else:
                    print('[ -- Output updated not supported for Test:',testrun.test.name,'-- ]')

    def verify(self):
        """ Update the status of all test runs """
        for testrun in self.testruns:
            if testrun.active:
                status, testrun.report = testrun.test.verify(testrun.output_path, testrun.exec_path)
                verifier_status = status[0]
                verifier_info = status[1]
                if verifier_status == 'pass':
                    testrun.status = _TestRunStatus.PASS
                    testrun.status_info = verifier_info
                elif verifier_status == 'fail':
                    testrun.status = _TestRunStatus.FAIL
                    testrun.status_info = verifier_info
                elif verifier_status == 'skip':
                    testrun.status = _TestRunStatus.SKIPPED
                    testrun.status_info = verifier_info
                else:
                    testrun.status = _TestRunStatus.FAIL
                    testrun.status_info = "Verifier returned unrecognized status and info: " + verifier_status + ", " + verifier_info
            else:
                testrun.status = _TestRunStatus.DEACTIVATED

    def _activate_tests_from_argument(self, test_subset_arg):
        test_subset_names = test_subset_arg.split(',')
        for testrun in self.testruns:
            testrun.active = False
        for name in test_subset_names:
            found = False
            for testrun in self.testruns:
                if name == testrun.test.name:
                    testrun.active = True
                    found = True
                    break
