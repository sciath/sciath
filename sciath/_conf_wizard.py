""" Interactive tool for populating Launcher configuration information """

import os
import shlex
import shutil

from sciath._sciath_io import py23input
from sciath._default_templates import _generate_default_template
from sciath.utility import DotDict

DEFAULTS_LOCAL = DotDict()
DEFAULTS_LOCAL.job_submission_command = ("sh",)
DEFAULTS_LOCAL.blocking = True
DEFAULTS_LOCAL.has_job_level_ranks = False

DEFAULTS_LSF = DotDict()
DEFAULTS_LSF.job_submission_command = (
    "sh",
    "-c",
    "bsub < $0",
)  # This allows "<".
DEFAULTS_LSF.blocking = False
DEFAULTS_LSF.has_job_level_ranks = True
DEFAULTS_LSF.mpi_launch = "mpirun"

DEFAULTS_SLURM = DotDict()
DEFAULTS_SLURM.job_submission_command = ("sbatch",)
DEFAULTS_SLURM.blocking = False
DEFAULTS_SLURM.has_job_level_ranks = True
DEFAULTS_SLURM.mpi_launch = "srun -n <ranks>"

DEFAULTS = {
    "local": DEFAULTS_LOCAL,
    "lsf": DEFAULTS_LSF,
    "slurm": DEFAULTS_SLURM,
}


def _query_user(description="",
                options=None,
                examples=None,
                default=None,
                validate=None):
    valid = False
    while not valid:
        options_string = "" if options is None else " <%s>" % ", ".join(options)

        print("%s%s" % (description, options_string))
        if examples is not None:
            print("Examples:")
            for example in examples:
                print(example)

        if default is not None:
            print("Hit enter to accept default [%s]" % default)

        result = py23input("> ")
        result = os.path.expandvars(result)
        result = os.path.expanduser(result)

        if result == "":
            result = default

        valid = options is None or result in options
        if validate is not None and valid:
            valid = validate(result)

    return result


def _launcher_interactive_configure(launcher):  #pylint: disable=too-many-statements, too-many-branches
    """ Create a new configuration file from user input """
    print("--------------------------------------------------------------")
    print("Creating new configuration file %s" % launcher.conf_filename)

    system_type_options = list(DEFAULTS.keys()) + ["generic"]
    system_type = _query_user("\n[?] Batch queuing system type",
                              options=system_type_options,
                              default="local")

    if system_type in DEFAULTS:
        launcher.job_submission_command = list(
            DEFAULTS[system_type].job_submission_command)
    else:
        launcher.job_submission_command = shlex.split(
            _query_user("\n[?] job submission command",
                        validate=lambda command: command is not None))

    if system_type in DEFAULTS:
        launcher.blocking = DEFAULTS[system_type].blocking
    else:
        launcher.blocking = _query_user("\n[?] blocking",
                                        options=["True", "False"]) == "True"

    if system_type in DEFAULTS:
        launcher.has_job_level_ranks = DEFAULTS[system_type].has_job_level_ranks
    else:
        launcher.has_job_level_ranks = _query_user(
            "\n[?] supports job-level ranks specification",
            options=["True", "False"]) == "True"

    mpi_launch_examples = [
        "  No MPI Required           : none",
        "  Local Machine (mpirun)    : mpirun -np <ranks>",
        "  Local Machine (mpiexec)   : mpiexec -np <ranks>",
        "  SLURM w/ aprun            : aprun -B",
        "  Native SLURM              : srun -n <ranks>",
        "  LSF (Euler)               : mpirun",
    ]
    petsc_dir = os.getenv("PETSC_DIR")
    petsc_arch = os.getenv("PETSC_ARCH")
    if petsc_dir and petsc_arch:
        mpi_launch_examples.append(
            "  Current PETSc MPI wrapper : %s -n <ranks>" %
            os.path.join(petsc_dir, petsc_arch, "bin", "mpiexec"))
    else:
        mpi_launch_examples.append(
            "  Example PETSc MPI wrapper : /users/myname/petsc/arch-xxx/bin/mpiexec -n <ranks>"
        )

    if system_type in DEFAULTS and "mpi_launch" in DEFAULTS[system_type]:
        mpi_launch_default = DEFAULTS[system_type].mpi_launch
    else:
        mpi_launch_default = "none"
    mpi_launch = _query_user(
        "\n[?] MPI launch command, including <ranks> for number of ranks",
        examples=mpi_launch_examples,
        default=mpi_launch_default,
        validate=launcher.is_valid_mpi_launch)

    launcher.set_mpi_launch(mpi_launch)

    launcher.account_name = _query_user("\n[?] Account to charge", default="")

    launcher.queue_name = _query_user("\n[?] Name of queue to submit tests to",
                                      default="")

    # Assume, dubiously, that defaults defined here have a default template defined,
    # that the user always wants to use.
    query_template = system_type not in DEFAULTS
    if query_template:
        template_source_filename = _query_user(
            "\n[?]Please select a file, to be copied here, to use as a Launcher template",
            validate=os.path.isfile,
        )
        template_filename = os.path.split(template_source_filename)[1]
        if not os.path.isfile(template_filename):
            shutil.copyfile(template_source_filename, template_filename)
        launcher.template_filename = template_filename
    else:
        launcher.template_filename = _generate_default_template(system_type)

    launcher.write_definition(launcher.conf_filename)

    print("")
    print("** The template for batch submission files is  %s" %
          launcher.template_filename)
    print("** You may modify it if desired, but it will be overwritten")
    print("** if you re-configure.")
    print("")
    print("** To change the config for your batch system, either")
    print("**  (i) delete the file %s  or" % launcher.conf_filename)
    print("** (ii) re-run with the command line arg --configure")
    print("--------------------------------------------------------------")
