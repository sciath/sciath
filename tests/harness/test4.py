#!/usr/bin/env python

# "Command line" use of the Harness, with no tests (add more with a file)

import sciath.harness

harness = sciath.harness.Harness()
harness.run_from_args()
