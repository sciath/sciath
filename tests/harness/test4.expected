[ -- Removing output for Test: foo -- ]
[ -- Removing output for Test: foo2 -- ]
[ -- Removing output for Test: foo_fail -- ]
[ -- Removing output for Test: missing -- ]
[ -- Removing output for Test: many_words -- ]
[36m[Executing foo][0m from <<PATH STRIPPED>>
  [cmd]  ['echo', 'foo']
[36m[Executing foo2][0m from <<PATH STRIPPED>>
  [cmd]  ['echo', 'foo']
[36m[Executing foo_fail][0m from <<PATH STRIPPED>>
  [cmd]  ['echo', 'foox']
[36m[Executing missing][0m from <<PATH STRIPPED>>
  [cmd]  ['echo', 'missing']
[36m[Executing many_words][0m from <<PATH STRIPPED>>
  [cmd]  ['echo', 'many words']
--- Error Reports ---
[Report for foo_fail]
1c1
< foo
---
> foox
[UnixDiff] Expected and output files are not identical
[UnixDiff] Expected file <<PATH STRIPPED>>"
[UnixDiff] Output file  : "sciath.job-foo_fail.stdout"
[Report for missing]
[UnixDiff] Expected file <<PATH STRIPPED>>" was not found
-- Summary --
foo : pass (verification was successful)
foo2 : pass (verification was successful)
foo_fail : fail (verification failed)
missing : fail (expected/comparison file not found)
many_words : pass (verification was successful)
TEST SUCCESS NOT CONFIRMED
